tools/codesign_mac_executable.py

```diff
@@ -0,0 +1,232 @@
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "Vorbis", "Mac", "libvorbis.dylib"),
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "PhysX3", "Mac", "libPhysX3.dylib"),
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "PhysX3", "Mac", "libPhysX3Common.dylib"),
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "PhysX3", "Mac", "libPxFoundation.dylib"),
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "PhysX3", "Mac", "libAPEX_Clothing.dylib"),
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "PhysX3", "Mac", "libPxPvdSDK.dylib"),
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "PhysX3", "Mac", "libNvCloth.dylib"),
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "PhysX3", "Mac", "libPhysX3Cooking.dylib"),
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "PhysX3", "Mac", "libAPEX_Legacy.dylib"),
+        os.path.join(args.macos_executable, "Contents", "UE4", "Engine", "Binaries", "ThirdParty", "PhysX3", "Mac", "libApexFramework.dylib"),
+        os.path.join(args.macos_executable, "Contents", "MacOS", os.path.splitext(executable_name)[0])]
+
+    assert os.path.exists(args.entitlements_file)
+
+    for file in files:
+        cmd = ["sudo", "codesign", "-f", "-s", "-v", "--options", "runtime", "--timestamp", "--entitlements", args.entitlements_file, "--sign", f"Developer ID Application: {args.developer_id}", file]
+        print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+        cmd_result = subprocess.run(cmd)
+        assert cmd_result.returncode == 0
+
+    os.makedirs(args.output_dir, exist_ok=True)
+
+    # create a zip file for notarization
+    temp_zip = os.path.join(args.output_dir, f"{os.path.splitext(os.path.basename(args.macos_executable))[0]}-temp.zip")
+    cmd = ["ditto", "-c", "-k", "--rsrc", "--keepParent", args.macos_executable, temp_zip]
+    print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+    cmd_result = subprocess.run(cmd)
+    assert cmd_result.returncode == 0
+
+    # send the zip file for notarization 
+    cmd = ["xcrun", "altool", "--notarize-app", "--primary-bundle-id", "org.embodiedaifoundation.spear", "--username", args.apple_username, "--password", args.apple_password, "--file", temp_zip]
+    print(f"[SPEAR | codesign_mac_executable.py] Executing: {' '.join(cmd)}")
+    ps = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+    output = subprocess.check_output(["grep", "RequestUUID = "], stdin=ps.stdout, text=True)
+    ps.wait()
+    ps.stdout.close()
+    request_uuid = " ".join(output[14:].split())
```

Comment:
why not split on "RequestUUID = ", similar to how you do in the build pak file script, instead of doing this brittle and not-self-documenting indexing trick [14:]

Reply:
Agreed, changed to `request_uuid = output.split("RequestUUID = ")[1].strip()` which is more clearer.

Reply:
Also, printing the piped output for more clarity.

Reply:
I tried to read `stdout` (something like 
```python 
ps = subprocess.Popen(cmd, stdout=subprocess.PIPE)
for line in ps.stdout:
    print(f"[SPEAR | sign_executable.py] {line.decode()}", end="")
output = subprocess.check_output(["grep", "     Status:"], stdin=ps.stdout, text=True)
```
, but this causes issues with the subsequent `check_output` cmd. If I try to read like this
```python 
ps = subprocess.Popen(cmd, stdout=subprocess.PIPE)
output = subprocess.check_output(["grep", "     Status:"], stdin=ps.stdout, text=True)
for line in ps.stdout:
    print(f"[SPEAR | sign_executable.py] {line.decode()}", end="")
```
 I'm not able to, i.e. no output. 

Not sure how to go about this. Any ideas?

Reply:
https://stackoverflow.com/questions/13398261/how-to-run-a-subprocess-with-python-wait-for-it-to-exit-and-get-the-full-stdout

Reply:
I'm not sure exactly how to do it, but it should definitely be possible to get the full output and the return code.

Reply:
I was able to solve this. Updated the code.